File: programs/bun-run.c
/*
 * bun-run: thin entrypoint that hands a Bun standalone executable to the
 * Kandelo `bun-run.js` bootstrap running on spidermonkey-node.
 *
 * `bun-run.js` (runtime/bun-run/bun-run.js, installed at
 * /usr/lib/kandelo/bun-run.js) spawns `bun-extract --prepare` to extract the
 * plaintext ESM module graph from the target Bun executable, installs a thin
 * `Bun` global shim, and natively `import()`s the cached entry point. This
 * program is just argv passthrough: it execs spidermonkey-node with the
 * bootstrap script and the caller's arguments, so `bun-run <binary> [args]`
 * behaves like running <binary> directly (e.g. as a future binfmt target).
 *
 * usage: bun-run <bun-executable> [args...]
 */
#include <unistd.h>
#include <string.h>

int main(int argc, char **argv) {
    /* bun-run <binary> [args...] -> spidermonkey-node /usr/lib/kandelo/bun-run.js <binary> [args...] */
    char *nargv[argc + 3];
    nargv[0] = "spidermonkey-node";
    nargv[1] = "/usr/lib/kandelo/bun-run.js";
    for (int i = 1; i < argc; i++) nargv[i + 1] = argv[i];
    nargv[argc + 1] = 0;
    execv("/usr/bin/spidermonkey-node", nargv);
    return 127;
}

File: runtime/bun-run/bun-run.js
"use strict";
// Kandelo bun-run bootstrap: extract (cached) + run a Bun executable on spidermonkey-node.
const cp = require("child_process");
const fs = require("fs");

const CACHE_ROOT = "/var/cache/kandelo/bun-run"; // volatile; never persisted (see spec)

function fail(msg, code) { process.stderr.write("bun-run: " + msg + "\n"); process.exit(code || 1); }

const binary = process.argv[2];
if (!binary) fail("usage: bun-run <bun-executable> [args...]", 2);
const appArgs = process.argv.slice(3);

// 1. Extract (cache-aware). One-shot; works with synchronous child_process today.
// NOTE: Kandelo's spidermonkey-node child_process shim ignores the `encoding`
// option and always hands back Buffers, so decode defensively with String()
// rather than assuming spawnSync gave us strings. It also never actually
// captures child stderr into result.stderr (always ''); `stdio: 'inherit'`
// for stderr makes the shim let bun-extract's own diagnostics (e.g. "not a
// Bun standalone executable...") flow straight through to our real stderr
// instead of being redirected to /dev/null, so a malformed-binary failure is
// not reduced to a bare "exit 1".
const r = cp.spawnSync("/usr/bin/bun-extract", ["--prepare", binary, CACHE_ROOT], {
  encoding: "utf8",
  stdio: ["ignore", "pipe", "inherit"],
});
const rStdout = String(r.stdout || "");
if (r.status !== 0) fail("extract failed (exit " + r.status + "); see stderr above", 1);
const entry = (rStdout.match(/^ENTRY=(.+)$/m) || [])[1];
if (!entry) fail("could not determine entry point (not a Bun executable?)", 1);

// 2. Thin Bun-global shim. Unimplemented members throw named errors (never silent).
const ni = (name) => () => { throw new Error("Bun." + name + " not implemented (Kandelo bun-run shim)"); };
globalThis.__breq = (id) =>
  require(typeof id === "string" && id.indexOf("file://") === 0 ? require("url").fileURLToPath(id) : id);
globalThis.Bun = {
  version: (process.versions && process.versions.node) || "0",
  isStandaloneExecutable: false,
  which(cmd) { try { const o = cp.spawnSync("sh", ["-lc", "command -v " + cmd], { encoding: "utf8" }); return String(o.stdout || "").trim() || null; } catch (_) { return null; } },
  file(p) { return { text: () => fs.promises.readFile(p, "utf8"), arrayBuffer: () => fs.promises.readFile(p), exists: () => Promise.resolve(fs.existsSync(p)) }; },
  write() { return ni("write")(); },
  spawn: ni("spawn"), serve: ni("serve"),
  stringWidth: (s) => String(s).length,
  stripANSI: (s) => String(s).replace(/\x1b\[[0-9;]*m/g, ""),
  gc() {}, deepEquals: (a, b) => JSON.stringify(a) === JSON.stringify(b),
};

// 3. Present the app with its own argv, then load it natively.
process.argv = [process.argv[0], entry].concat(appArgs);
import(entry).catch((e) => fail((e && (e.stack || e.message)) || String(e), 1));
